hpy/devel/src/runtime/hlist.c

```diff
@@ -0,0 +1,135 @@
+/**
+ * A manager for a list of HPy handles, allowing handles to be tracked
+ * and closed as a group.
+ *
+ * Note::
+ *    HList always keeps space for one extra handle free so that
+ *    HList_Track can always store the handle being passed to it,
+ *    even if it fails to automatically create space for future
+ *    handles. This allows HList_Free to close all handles passed to
+ *    HList_Track.
+ *
+ *    As a consequence, the minimum size for an HList is one handle.
+ *
+ * Example usage (inside an HPyDef_METH function)::
+ *
+ * long i;
+ * HPy key, value;
+ * HList hl;
+ *
+ * hl = HList_New(ctx);  // track the key-value pairs
+ * if (hl == NULL)
+ *     return HPy_NULL;
+ *
+ * HPy dict = HPyDict_New(ctx);
+ * if (HPy_IsNull(dict))
+ *     goto error;
+ *
+ * for (i=0; i<5; i++) {
+ *     key = HPyLong_FromLong(ctx, i);
+ *     if (HPy_IsNull(key))
+ *         goto error;
+ *     if (HList_Track(ctx, hl, key) < 0)
+ *         goto error;
+ *     value = HPyLong_FromLong(ctx, i * i);
+ *     if (HPy_IsNull(value)) {
+ *         goto error;
+ *     }
+ *     if (HList_Track(ctx, hl, value) < 0)
+ *         goto error;
+ *     result = HPy_SetItem(ctx, dict, key, value);
+ *     if (result < 0)
+ *         goto error;
+ * }
+ *
+ * success:
+ *    HList_Free(ctx, hl);
+ *    return dict;
+ *
+ * error:
+ *    HList_Free(ctx, hl);
+ *    HPy_Close(ctx, dict);
+ *    HPyErr_SetString(ctx, ctx->h_ValueError, "Failed!");
+ *    return HPy_NULL;
+ */
+
+#include <Python.h>
+#include "hpy.h"
+
+#define _HLIST_INITIAL_SIZE (5 + 1)
+
+struct _HList_s {
+    HPy_ssize_t size;
+    HPy_ssize_t next;
+    HPy *handles;
+};
+
+
+HPyAPI_RUNTIME_FUNC(HList)
+HList_New(HPyContext ctx)
+{
+    HList hl;
+    hl = PyMem_Malloc(sizeof(struct _HList_s));
+    if (hl == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+    hl->handles = PyMem_Calloc(_HLIST_INITIAL_SIZE, sizeof(HPy));
+    if (hl->handles == NULL) {
+        PyMem_Free(hl);
+        PyErr_NoMemory();
+        return NULL;
+    }
+    hl->size = _HLIST_INITIAL_SIZE;
+    hl->next = 0;
+    return hl;
+}
+
+HPyAPI_RUNTIME_FUNC(int)
+HList_Resize(HPyContext ctx, HList hl, HPy_ssize_t size)
```

Comment:
does it need to be a public function? It looks to me that the only valid/reasonable use case it to call it from within `HList_Track`?

Reply:
Agreed about the name. I like HPyTracker.

Re Python.h -- note that, e.g., ctx_listbuilder in runtime also includes Python.h. Are those different somehow? Should they not use Python.h? Or not be in runtime? Or something else?

Reply:
I felt that it would be very helpful to expose it because if one can calculate the number of handles needed, then one can avoid a few mallocs by calling Resize explicitly oneself.

Reply:
That said, I exposed Resize before the extra handle logic was added, which perhaps makes Resize currently a bit too complex to use well. Perhaps Resize should internally allocate space for one more handle than requested.

Reply:
another alternative is to offer only `HPyTracker_New(ctx, initial_size)`. If you know it in advance, you pass a precise number, else you pass 0 (or -1, if you want a default) and let `HPyTracker` to take care of that.